data/src/main/java/com/linkedin/data/codec/entitystream/ProtobufDataDecoder.java

```diff
@@ -0,0 +1,891 @@
+  }
+
+  class ProtobufStreamDataParser implements NonBlockingDataParser
+  {
+    private final SymbolTable _symbolTable;
+
+    private final Deque<Integer> _complexObjTokenSizeStack = new ArrayDeque<>();
+    private int _currComplexObjTokenSize = -1;
+
+    private byte[] _input;  //holds feed input bytes
+    private int _limit;
+    private int _pos;
+
+    private boolean _eofInput;  //no more inputs can be feed if this is set to true
+
+    private final TextBuffer _textBuffer;  //buffer to hold parsed string characters.
+    private int _bufferPos = -1;  //signify no. of chars in text buffers as buffer is reused to avoid thrashing
+
+    private int _pendingCharUtfRep;  // no. of bytes used by Utf-8 multi-byte representation of pending char
+    private int _pendingIntShifts = -1;  // remaining bits/bytes for int32/64
+    private long _pendingInt64;
+    private int _pendingInt32;
+
+    // Stores current token returned from #nextToken else Token#NOT_AVAILABLE
+    private Token _currentToken;
+    private byte _currentOrdinal = -1;
+
+    //Below value variables hold parsed value for current token returned from #nextToken
+    private byte[] _bytesValue;
+    private String _stringValue;
+    private int _intValue;
+    private long _longValue;
+
+    ProtobufStreamDataParser(SymbolTable symbolTable)
+    {
+      _symbolTable = symbolTable == null ? EmptySymbolTable.SHARED : symbolTable;
+      _textBuffer = new TextBuffer(ProtoReader.DEFAULT_TEXT_BUFFER_SIZE);
+    }
+
+    @Override
+    public void feedInput(byte[] data, int offset, int len) throws IOException
+    {
+      if (data == null || data.length < offset + len)
+      {
+        throw new IllegalArgumentException("Bad arguments");
+      }
+
+      if (_pos >= _limit && !_eofInput)
+      {
+        _pos = offset;
+        _limit = offset + len;
+        _input = data;
+      }
+      else
+      {
+        throw new IOException("Invalid state: Parser cannot accept more data");
+      }
+    }
+
+    @Override
+    public void endOfInput()
+    {
+      _eofInput = true;
+    }
+
+    @Override
+    public Token nextToken() throws IOException
+    {
+      // First: regardless of where we really are, need at least one more byte;
+      // can simplify some of the checks by short-circuiting right away
+      if (_pos >= _limit) {
+        Token endComplexObjToken = readEndComplexObj();
+        if (endComplexObjToken != NOT_AVAILABLE)
+        {
+          return finishToken(endComplexObjToken);
+        }
+        if (_eofInput) {
+          return EOF_INPUT;
+        }
+        return NOT_AVAILABLE;
+      }
+      if (_currentToken != NOT_AVAILABLE)
+      {
+        _currentToken = readEndComplexObj();
+        if (_currentToken != NOT_AVAILABLE)
+        {
+          return finishToken(_currentToken);
+        }
+        _currentOrdinal = _input[_pos++];
+        //release bytes array if previous token was Token#RAW_BYTES
+        _bytesValue = null;
```

Comment:
I was using _bytesValue as buffer too, to avoid references long byte arrays but now on 2nd thoughts it won't matter much.
Added a byteBuffer separately which can be reset similar to other read pending objects.